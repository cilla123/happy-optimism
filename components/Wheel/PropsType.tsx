export interface BaseWheelProps {
  value?: string | number;
  defaultValue?: string | number;
  valueMember?: string;
  dataSource?: object[];
  onChange?: (value?: any) => void;
  itemRender?: (item?: object) => string;
  disabled?: boolean;
  onTransition?: (value: boolean) => void;
}
