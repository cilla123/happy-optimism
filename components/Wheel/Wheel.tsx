import React, { Component } from 'react';
import classnames from 'classnames';
import BScroll from 'better-scroll';
import { BaseWheelProps } from './PropsType';
import { isArray } from '../utils/validate';
import './findIndex.js';

const getValue = (props, defaultValue?: any) => {
  if ('value' in props) {
    return props.value;
  }
  if ('defaultValue' in props) {
    return props.defaultValue;
  }
  if (isArray(props.dataSource) && props.dataSource[0]) {
    return props.dataSource[0][props.valueMember];
  }
  return defaultValue;
};

export interface WheelProps extends BaseWheelProps {
  prefixCls?: string;
  className?: string;
}

export default class Wheel extends Component<WheelProps, any> {

  static defaultProps = {
    prefixCls: 'za-wheel',
    dataSource: [],
    valueMember: 'value',
    itemRender: item => item.label,
  };

  private BScroll;
  private wrapper;

  constructor(props) {
    super(props);
    this.state = {
      value: getValue(props),
    };

  }

  componentDidMount() {
    const { prefixCls } = this.props;
    this.BScroll = new BScroll(this.wrapper , {
      wheel: {
        selectedIndex: 0,
        wheelWrapperClass: `${prefixCls}-content`,
        wheelItemClass: `${prefixCls}-item`,
        adjustTime: 100,
      },
      probeType: 3,
    });

    if (this.props.disabled) {
      this.BScroll.disable();
    }

    const initIndex = this.BScroll.getSelectedIndex(this.state.value);
    this.BScroll.wheelTo(initIndex);

    this.BScroll.on('scroll', () => {
      this.props.onTransition!(true);
    });

    this.BScroll.on('scrollEnd', () => {
      const { dataSource, valueMember } = this.props;
      const index = this.BScroll.getSelectedIndex();
      const child = dataSource![index];
      if (child) {
        this.fireValueChange(child[valueMember!]);
        this.props.onTransition!(this.BScroll.isInTransition);
      } else if (console.warn) {
        console.warn('child not found', dataSource, index);
      }
    });
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      this.setState({
        value: nextProps.value,
      });
    }

    if (nextProps.disabled) {
      this.BScroll.disable();
    }
    const newIndex = this.getSelectedIndex(nextProps.value, nextProps.dataSource);
    this.BScroll.wheelTo(newIndex);

  }

  componentDidUpdate() {
    this.BScroll.refresh();
  }

  componentWillUnmount() {
    this.BScroll.destroy();
  }

  fireValueChange = (value) => {
    if (value === this.state.value) {
      return;
    }

    if (!('value' in this.props)) {
      this.setState({ value });
    }

    const { onChange } = this.props;
    if (typeof onChange === 'function') {
      onChange(value);
    }
  }

  getSelectedIndex(value, dataSource) {
    const { valueMember } = this.props;
    return dataSource!.findIndex((item) => item[valueMember!] === value);
  }

  render() {
    const {
      prefixCls,
      className,
      valueMember,
      dataSource,
      itemRender,
    } = this.props;

    const { value } = this.state;
    const items = dataSource!.map((item, index) => {
      const itemCls = classnames(`${prefixCls}-item`, {
        [`${prefixCls}-item-selected`]: value === item[valueMember!],
      });

      return (
        <div
          key={+index}
          className={itemCls}
        >
          {itemRender!(item)}
        </div>
      );
    });

    const rollerCls = classnames(`${prefixCls}`, className);

    return (
      <div className={rollerCls} ref={(wrapper) => { this.wrapper = wrapper; }}>
        {/* <div className={`${prefixCls}-indicator`} ref={(indicator) => { this.indicator = indicator; }} /> */}
        <div className={`${prefixCls}-content`}>
          {items}
        </div>
      </div>
    );
  }
}
