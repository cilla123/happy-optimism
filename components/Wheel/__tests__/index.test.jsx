import React from 'react';
import { mount } from 'enzyme';
import toJson from 'enzyme-to-json';
import Wheel from '../index';

describe('Wheel', () => {
  it('Wheel render visible', () => {
    // jest.useFakeTimers();
    const wrapper = mount(
      <Wheel
        dataSource={[
          { value: '1', label: '选项一' },
          { value: '2', label: '选项二' },
        ]}
        defaultValue="1"
        value="1"
        />
    );
    expect(toJson(wrapper)).toMatchSnapshot();
    wrapper.setProps({ value: '2' });
    // jest.runAllTimers();
    wrapper.unmount();
  });
});
