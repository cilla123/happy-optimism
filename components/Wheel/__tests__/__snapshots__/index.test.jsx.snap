// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Wheel Wheel render visible 1`] = `
<Wheel
  dataSource={
    Array [
      Object {
        "label": "选项一",
        "value": "1",
      },
      Object {
        "label": "选项二",
        "value": "2",
      },
    ]
  }
  defaultValue="1"
  itemRender={[Function]}
  prefixCls="za-wheel"
  value="1"
  valueMember="value"
>
  <div
    className="za-wheel"
  >
    <div
      className="za-wheel-content"
    >
      <div
        className="za-wheel-item za-wheel-item-selected"
        key="0"
      >
        选项一
      </div>
      <div
        className="za-wheel-item"
        key="1"
      >
        选项二
      </div>
    </div>
  </div>
</Wheel>
`;
